include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,835 @@
+            // TODO(achal): Return Iterator&?
+            // inline void operator++()
+            // {
+            //     m_cmd = reinterpret_cast<ICommand*>(reinterpret_cast<uint8_t*>(m_cmd) + m_size);
+            //     m_size = ;
+            // }
         };
 
-        IGPUCommandPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::bitflag<E_CREATE_FLAGS> _flags, uint32_t _familyIx) : IBackendObject(std::move(dev)), m_flags(_flags), m_familyIx(_familyIx) {}
+        CCommandSegment()
+        {
+            header.m_commandAllocator = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, COMMAND_SEGMENT_ALIGNMENT, STORAGE_SIZE);
+            header.m_next = nullptr;
 
-        core::bitflag<E_CREATE_FLAGS> getCreationFlags() const { return m_flags; }
-        uint32_t getQueueFamilyIndex() const { return m_familyIx; }
+            wipeNextCommandSize();
+        }
 
-        // OpenGL: nullptr, because commandpool doesn't exist in GL (we might expose the internal allocator in the future)
-        // Vulkan: const VkCommandPool*
-        virtual const void* getNativeHandle() const = 0;
+        template <typename Cmd, typename... Args>
+        void* allocate(const Args&... args)
+        {
+            const uint32_t cmdSize = Cmd::calc_size(args...);
+            const auto address = header.m_commandAllocator.alloc_addr(cmdSize, alignof(Cmd));
+            if (address == decltype(header.m_commandAllocator)::invalid_address)
+                return nullptr;
 
-    protected:
-        virtual ~IGPUCommandPool() = default;
+            wipeNextCommandSize();
+
+            void* cmdMem = m_data + address;
+            return cmdMem;
+        }
+
+        inline void setNext(CCommandSegment* segment) { header.m_next = segment; }
+        inline CCommandSegment* getNext() const { return header.m_next; }
+        inline ICommand* getFirstCommand() { return reinterpret_cast<ICommand*>(m_data); } // TODO(achal): Probably remove?
+        inline uint8_t* getData() { return m_data; }
+
+        inline CCommandSegment::Iterator begin()
+        {
+            ICommand* cmd = reinterpret_cast<ICommand*>(m_data);
+            // This assumes that cmd will always be a valid ICommand.
+            // return { cmd, cmd->getSize() };
+            return { cmd };
+        }
+
+        inline CCommandSegment::Iterator end()
+        {
+            return { reinterpret_cast<ICommand*>(m_data + header.m_commandAllocator.get_allocated_size()) };
+        }
+
+    private:
+        alignas(ICommand) uint8_t m_data[STORAGE_SIZE];
+
+        void wipeNextCommandSize()
+        {
+            const auto cursor = header.m_commandAllocator.get_allocated_size();
+            const uint32_t wipeSize = offsetof(IGPUCommandPool::ICommand, m_size) + sizeof(IGPUCommandPool::ICommand::m_size);
+            if (cursor + wipeSize < header.m_commandAllocator.get_total_size())
```

Comment:
`wipeSize` is misleadingly named now, also the comparison should have been a `<=` not a `<`

If I may suggest a semantically clearer rewrite of  `wipeNextCommandSize`
```cpp
            const auto nextCmdOffset = header.m_commandAllocator.get_allocated_size();
            const auto wipeEnd = nextCmdOffset+offsetof(ICommand,m_size)+sizeof(ICommand::m_size);
            if (wipeEnd<=STORAGE_SIZE)
                reinterpret_cast<ICommand*>(m_data+nextCmdOffset)->m_size = 0;
```

Reply:
Why risk possible UB with this,
```C++
reinterpret_cast<ICommand*>(m_data+nextCmdOffset)->m_size = 0;
```
if we can just use `memset`?

Reply:
This will also require us to make `m_size` `public` (or declare friendship) and also make it non-`const`.

Reply:
Anyway, resolved it the exactly the way you suggested.